### PR TITLE
fix: remove null checks before free. Reverses commit eb5395b

### PR DESCRIPTION
**Pull request by lemire**
See commit https://github.com/RoaringBitmap/CRoaring/commit/eb5395b649fc15ae8fb6058c2b46cbec6a16cb51 where I added these checks.

Might fix https://github.com/RoaringBitmap/CRoaring/issues/345

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -303,6 +303,20 @@ int main(){
...
+    .malloc = malloc,
+    .realloc = realloc,
+    .calloc = calloc,
+    .free = free,
+    .aligned_malloc = roaring_bitmap_aligned_malloc,
+    .aligned_free = roaring_bitmap_aligned_free,
+};
+```
+
+We use the C convention regarding `free`: `free(NULL)` should have no effect. Thus our code does not check whether the provided pointer is non-null.
```
Comment (Dr-Emann) [inline README.md:318]: I think I'd word this as something like "we require the .free/.aligned_free function to follow the C convention where free(NULL) has no effect", stating it as a requirement vs saying what we do and having the user have to figure out what it means they need to do.

File: src/containers/run.c [hunk 2]
```diff
@@ -189,11 +189,8 @@ void run_container_offset(const run_container_t *c, container_t **loc,
 
 /* Free memory. */
 void run_container_free(run_container_t *run) {
-    if (run->runs !=
-        NULL) {  // Jon Strabala reports that some tools complain otherwise
-        roaring_free(run->runs);
-        run->runs = NULL;  // pedantic
-    }
+    roaring_free(run->runs);
+    run->runs = NULL;  // pedantic
```
Comment (Dr-Emann) [inline src/containers/run.c:193]: I think especially in these cases where the next line (in a 3 line function) clearly means the value set here doesn't matter, we shouldn't bother setting it to null.

Same above in array_container_free

**Review by Dr-Emann — APPROVED** [review]
Some nits, but would be okay with going in as is.

File: src/containers/run.c (hunk 2 above)
Comment (lemire) [reply src/containers/run.c:193]: I thought I had removed the pedantic settings. Let me solve this.

Comment (lemire) [reply src/containers/run.c]: I pushed. Resolved.

File: README.md (hunk 1 above)
Comment (lemire) [reply README.md]: Wording updated.

**Review by Dr-Emann — APPROVED** [review]

File: src/containers/array.c [hunk 3]
```diff
@@ -145,11 +145,7 @@ int array_container_shrink_to_fit(array_container_t *src) {
 
 /* Free memory. */
 void array_container_free(array_container_t *arr) {
-    if (arr->array !=
-        NULL) {  // Jon Strabala reports that some tools complain otherwise
-        roaring_free(arr->array);
-        arr->array = NULL;  // pedantic
-    }
+    roaring_free(arr->array);
```
Comment (Oppen) [inline src/containers/array.c:148]: I believe we should still check that `arr != NULL` before accessing `arr->array`.

File: src/containers/bitset.c [hunk 4]
```diff
@@ -130,11 +130,7 @@ void bitset_container_add_from_range(bitset_container_t *bitset, uint32_t min,
 
 /* Free memory. */
 void bitset_container_free(bitset_container_t *bitset) {
-    if (bitset->words !=
-        NULL) {  // Jon Strabala reports that some tools complain otherwise
-        roaring_aligned_free(bitset->words);
-        bitset->words = NULL;  // pedantic
-    }
+    roaring_aligned_free(bitset->words);
```
Comment (Oppen) [inline src/containers/bitset.c:133]: And here as well, right?

File: src/containers/run.c [hunk 5]
```diff
@@ -189,11 +189,7 @@ void run_container_offset(const run_container_t *c, container_t **loc,
 
 /* Free memory. */
 void run_container_free(run_container_t *run) {
-    if (run->runs !=
-        NULL) {  // Jon Strabala reports that some tools complain otherwise
-        roaring_free(run->runs);
-        run->runs = NULL;  // pedantic
-    }
+    roaring_free(run->runs);
```
Comment (Oppen) [inline src/containers/run.c:192]: And here.

Comment (lemire) [reply src/containers/run.c:193]: added back the check.

File: src/containers/bitset.c (hunk 4 above)
Comment (lemire) [reply src/containers/bitset.c:134]: added back the check.

File: src/containers/array.c (hunk 3 above)
Comment (lemire) [reply src/containers/array.c:149]: added back the check.

**Comment by lemire** [comment]
@Oppen @Dr-Emann In the end, we come back almost full circle and we still have the null checks in all but one case.

**Comment by lemire** [comment]
I am going to merge this because it is cleaner code.

**Pull request merged**